### PR TITLE
Fix privileged mechanism - [MOD-6838]

### PR DESCRIPTION
**Pull request by GuyAv46**
### Describe the changes in the pull request

Introduce a ticketing mechanism to the thread pool to correctly ensure we have enough threads performing as privileged when choosing a queue to pull a job from.

#### Example of a problematic scenario:
3 threads, 1 should be privileged, infinite jobs (both high and low priority)
##### Before

1. All the threads are working (many jobs to execute)
2. A thread finishes his current job and is about to pull another one
3. Since `num_jobs_in_progress` is higher than the number of expected privilege threads (2 > 1), the thread behaves as if it is not privileged
4. This will repeat for all the threads. None of the threads behave as privileged!
##### After

1. All the threads are working (many jobs to execute)
2. A thread finishes his current job and is about to pull another one
3. If one of the other threads is privileged, it will behave as unprivileged and take a high or low priority job (50:50).
    If both others are not privileged, a privilege ticket will be available, so the thread will take it and behave as privileged.

#### Mark if applicable

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: tests/cpptests/test_cpp_thpool.cpp [hunk 1]
```diff
@@ -207,3 +207,84 @@ TEST_P(PriorityThpoolTestFunctionality, TestPauseResume) {
...
+     * 3. We add a state check job for each priority.
+     * 4. Add a final job to release the other waiting thread (low priority job)
+     * 5. Release one of the threads from the waiting job to execute the rest of the jobs.
+     *
+     * We expect that the unblocked thread will understand that it is privileged and will take the high priority jobs
+     * before the low priority jobs. therefore, the high priority state check is expected to see 5 high priority jobs
+     * and 0 low priority jobs, and the low priority state check is expected to see 5 high priority jobs and 5 low priority jobs.
+     *
+     * Before the mechanism fix, the unblocked thread would see that there is one running thread, and would assume that it is
+     * unprivileged (num jobs in progress is not smaller than the number of privileged threads), and would take once a high priority
```
Comment (meiravgri) [inline tests/cpptests/test_cpp_thpool.cpp:261]: wasn't clear to me if "it" refers to the blocked or to the unblocked thread. 
Feel free to rephrase 
```suggestion
     * Before the mechanism fix, the unblocked thread would assume that it is unprivileged if it sees that there is one running thread
     * (num jobs in progress is not smaller than the number of privileged threads), and would take once a high priority
```

**Review by meiravgri — COMMENTED** [review]
Well done! 
The test is very clean and clear :)
I suggest to run some mixed-load benchmarks before we insert this fix (I'll give you the script)

File: tests/cpptests/test_cpp_thpool.cpp [hunk 2]
```diff
@@ -207,3 +207,84 @@ TEST_P(PriorityThpoolTestFunctionality, TestPauseResume) {
...
 }
+
+/* ========================== NUM_THREADS = 2, NUM_PRIVILEGED = 1 ========================== */
+THPOOL_TEST_SUITE(PriorityThpoolTestPrivilegedAndUnprivileged, 2, 1)
+
+TEST_P(PriorityThpoolTestPrivilegedAndUnprivileged, TestTakingAsPrivileged) {
+    size_t num_jobs = 5;
+
+    // This job will keep the thread busy until we tell it to finish.
+    auto waitForSignFunc = [](void *p) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/cpptests/test_cpp_thpool.cpp:218]: Yes 🥲

**Review by meiravgri — DISMISSED** [review]

File: deps/thpool/thpool.c [hunk 3]
```diff
@@ -721,14 +727,17 @@ static priorityJobCtx priority_queue_pull(priorityJobqueue *priority_queue_p) {
 
   if (!job_p) {
     is_admin = false;
-    if (priority_queue_p->num_jobs_in_progress <
-        priority_queue_p->n_privileged_threads) {
+    // When taking a privilege ticket, we must hold the lock (read-and-then-update not atomic)
+    if (priority_queue_p->privilege_tickets > 0) {
       /* This is a privileged thread id, try taking from the high priority
```
Comment (alonre24) [inline deps/thpool/thpool.c:732]:
```suggestion
      /* Reserved for high priority tasks - try taking from the high priority
```

File: deps/thpool/thpool.c [hunk 4]
```diff
@@ -91,6 +92,8 @@ typedef struct priority_queue {
                                        from queue */
   unsigned char n_privileged_threads; /* number of threads that always run high
                                          priority tasks */
+  atomic_uchar privilege_tickets; /* number of currently available privilege
```
Comment (alonre24) [inline deps/thpool/thpool.c:95]: I suggest changing the terminology here, as these are no longer privileged threads, but rather threads reserved for high-priority tasks. Maybe `num_resreved_threads`

File: deps/thpool/thpool.c [hunk 5]
```diff
@@ -59,6 +59,7 @@ typedef struct job {
 typedef struct {
   job *job;
   bool is_admin;
+  bool is_privileged;
```
Comment (alonre24) [inline deps/thpool/thpool.c:62]:
```suggestion
  bool reserved_for_high_priority;
```
or something like that

**Review by alonre24 — COMMENTED** [review]
Looks good, beautiful test

File: deps/thpool/thpool.c (hunk 4 above)
Comment (meiravgri) [reply deps/thpool/thpool.c:95]: `num_high_priority_threads` ?

**Pull request merged**